zipline/data/fx/base.py

```diff
@@ -36,3 +39,34 @@ def get_rates(self, rate, quote, bases, dts):
             The row at index i corresponds to the dt in dts[i].
             The column at index j corresponds to the base currency in bases[j].
         """
+
+    @default
+    def get_rate_scalar(self, rate, quote, base, dt):
+        """Scalar version of ``get_rates``.
+
+        Parameters
+        ----------
+        rate : str
+            Rate type to load. Readers intended for use with the Pipeline API
+            should support at least ``zipline.data.fx.DEFAULT_FX_RATE``, which
+            will be used by default for Pipeline API terms that don't specify a
+            specific rate.
+        quote : str
+            Currency code of the currency to convert into.
+        base : str
+            Currency code of the currency to convert from.
+        dt : np.datetime64 or pd.Timestamp
+            Datetime on which to load rate.
+
+        Returns
+        -------
+        rate : np.float64
+            Exchange rate from base -> quote on dt.
+        """
+        rates_array = self.get_rates(
+            rate,
+            quote,
+            bases=np.array([base], dtype=object),
+            dts=pd.DatetimeIndex([dt], tz='UTC'),
```

Comment:
I'm not seeing much in the Pandas docs about what the `tz` argument to `DatetimeIndex` actually does -- is it just used to convert its inputs? If so, should we document its type as "convertible to datetime64", or do you want to limit what this interface officially supports?

Also, some failure modes we may or may not want to account for:

- Someone in Japan provides `dt='2020-01-01'` (or a timezone-naive Timestamp or datetime64 object, intended to represent local time) and gets results for a different market day than they expected.
- They try again with `dt=pd.Timestamp('2020-01-01', tz='Japan')` and get`TypeError: Already tz-aware, use tz_convert to convert.`

Not sure what happens (or should happen) if non-UTC timezone-aware datetimes are passed directly to `get_rates`. Looks like the `test_fx` tests all use UTC Timestamps: should that just be the requirement in the interface?

Reply:
>  If so, should we document its type as "convertible to datetime64", or do you want to limit what this interface officially supports?

In general, `DatetimeIndex` will accept anything that pandas can convert to a datetime. For example:

```python
In [8]: pd.DatetimeIndex(['2014-01-02', 5])
Out[8]: DatetimeIndex(['2014-01-02 00:00:00', '1970-01-01 00:00:00.000000005'], dtype='datetime64[ns]', freq=None)
```

So, if the goal here is to be maximally accurate, the signature of this is probably "anything that pandas can convert to a datetime". In practice, however, this class is a relatively low-level interface. I expect its primary clients to be pipeline loaders, not end users, so I'd rather advertise the interface I actually expect, which is Timestamp or datetime64.

> Looks like the test_fx tests all use UTC Timestamps: should that just be the requirement in the interface?

One of the relatively bad mistakes we made in Zipline a long time ago is that we uniformly represent "dates" as pandas `Timestamp`s at midnight, localized to utc, of the represented date. This is an objectively bad representation for a date for a variety of reasons, but it's been our representation for a long time, and it's pretty deeply ingrained into the codebase. That's the expected representation here, and I updated the docs on `get_rates` to reflect that here: https://github.com/quantopian/zipline/pull/2613/commits/8a361e7613984645f4d5bab1a6fff2b1bea48f0b.

> Someone in Japan provides dt='2020-01-01' (or a timezone-naive Timestamp or datetime64 object, intended to represent local time) and gets results for a different market day than they expected.

Both of these cases would result in loading the rates for the date requested, which is probably the expected behavior.

> They try again with dt=pd.Timestamp('2020-01-01', tz='Japan') and getTypeError: Already tz-aware, use tz_convert to convert.

As you note, this would be an error, which I think is reasonable. We never expect to see non-utc-localized dates in this function (or, really, anywhere in zipline).